core/chains/evm/logpoller/log_poller.go

```diff
@@ -0,0 +1,430 @@
+		lggr:              lggr,
+		replay:            make(chan int64),
+		done:              make(chan struct{}),
+		pollPeriod:        pollPeriod,
+		finalityDepth:     finalityDepth,
+		backfillBatchSize: backfillBatchSize,
+		addresses:         make(map[common.Address]struct{}),
+		topics:            make(map[int]map[common.Hash]struct{}),
+	}
+}
+
+// MergeFilter will update the filter with the new topics and addresses.
+// Clients may chose to MergeFilter and then replay in order to ensure desired logs are present.
+func (lp *LogPoller) MergeFilter(topics []common.Hash, address common.Address) {
+	lp.filterMu.Lock()
+	defer lp.filterMu.Unlock()
+	lp.addresses[address] = struct{}{}
+	// [[A, B], [C]] + [[D], [], [E]] = [[A, B, D], [C], [E]]
+	for i := 0; i < len(topics); i++ {
+		if lp.topics[i] == nil {
+			lp.topics[i] = make(map[common.Hash]struct{})
+		}
+		lp.topics[i][topics[i]] = struct{}{}
+	}
+}
+
+func (lp *LogPoller) filterAddresses() []common.Address {
+	lp.filterMu.Lock()
+	defer lp.filterMu.Unlock()
+	var addresses []common.Address
+	for addr := range lp.addresses {
+		addresses = append(addresses, addr)
+	}
+	sort.Slice(addresses, func(i, j int) bool {
+		return bytes.Compare(addresses[i][:], addresses[j][:]) < 0
+	})
+	return addresses
+}
+
+func (lp *LogPoller) filterTopics() [][]common.Hash {
+	lp.filterMu.Lock()
+	defer lp.filterMu.Unlock()
+	var topics [][]common.Hash
+	for idx := 0; idx < len(lp.topics); idx++ {
+		var topicPosition []common.Hash
+		for topic := range lp.topics[idx] {
+			topicPosition = append(topicPosition, topic)
+		}
+		sort.Slice(topicPosition, func(i, j int) bool {
+			return bytes.Compare(topicPosition[i][:], topicPosition[j][:]) < 0
+		})
+		topics = append(topics, topicPosition)
+	}
+	return topics
+}
+
+// Replay signals that the poller should resume from a new block.
+// Blocks until the replay starts.
+func (lp *LogPoller) Replay(ctx context.Context, fromBlock int64) error {
+	latest, err := lp.ec.BlockByNumber(ctx, nil)
+	if err != nil {
+		return err
+	}
+	if fromBlock < 1 || uint64(fromBlock) > latest.NumberU64() {
+		return errors.Errorf("Invalid replay block number %v, acceptable range [1, %v]", fromBlock, latest.NumberU64())
+	}
+	lp.replay <- fromBlock
+	return nil
+}
+
+func (lp *LogPoller) Start(parentCtx context.Context) error {
+	return lp.StartOnce("LogPoller", func() error {
+		ctx, cancel := context.WithCancel(parentCtx)
+		lp.ctx = ctx
+		lp.cancel = cancel
+		go lp.run()
+		return nil
+	})
+}
+
+func (lp *LogPoller) Close() error {
+	return lp.StopOnce("LogPoller", func() error {
+		lp.cancel()
+		<-lp.done
+		return nil
+	})
+}
+
+func (lp *LogPoller) run() {
+	defer close(lp.done)
+	tick := time.After(0)
+	var start int64
+	for {
+		select {
+		case <-lp.ctx.Done():
+			return
+		case fromBlock := <-lp.replay:
+			lp.lggr.Warnw("Replay requested", "from", fromBlock)
+			start = fromBlock
+		case <-tick:
+			tick = time.After(utils.WithJitter(lp.pollPeriod))
+			if start == 0 {
```

Comment:
nit: would prefer to see the logic on this branch separated in a different case with its own tick, then you would have a tick to start and a tick to poll, like so:

```go
var regularTick <-chan time.Time
startTick := time.After(0)
for {
  select {
    case <-ctx.Done():
      return
    case <-startTick:
      // the code on this start == 0 branch
      if success { startTick = nil; regularTick = time.After(0); }
    case <-regularTick:
      regularTick = time.After(pollPeriod);
      // the rest
  }
}
```